pmd-core/src/main/java/net/sourceforge/pmd/benchmark/TimeTracker.java

```diff
@@ -0,0 +1,240 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.benchmark;
+
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * A time tracker class to measure time spent on different sections of PMD analysis.
+ * The class is thread-aware, allowing to differentiate CPU and wall clock time.
+ * 
+ * @author Juan Martín Sotuyo Dodero
+ */
+public final class TimeTracker {
```

Comment:
Maybe this should be named `TimeTrackingUtil` to make the `Util` part explicit? Your call

Reply:
Mmmm... I don't know, this isn't really a utility class, it has internal state and all. It's more of a singleton, but without the overhead of asking for the instance each time...